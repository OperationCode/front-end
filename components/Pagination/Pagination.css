.Pagination {
}

.controls {
  cursor: pointer;
}

.pageNumberActive {
  color: red;
}
