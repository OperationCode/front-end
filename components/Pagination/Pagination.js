import React from 'react';
import { number, string, object } from 'prop-types';
import LeftAngleIcon from 'static/images/icons/FontAwesome/angle-left-solid.svg';
import RightAngleIcon from 'static/images/icons/FontAwesome/angle-right-solid.svg';
import { PREV_PAGE_BUTTON, NEXT_PAGE_BUTTON } from '../../common/constants/testIDs';
import PaginationItem from './PaginationItem/PaginationItem';
import styles from './Pagination.module.css';

Pagination.propTypes = {
  currentPage: number.isRequired,
  pathname: string.isRequired,
  query: object.isRequired,
  totalPages: number.isRequired,
};

export const developmentErrors = {
  currentPageValue: value => `The value passed for currentPage is ${value}.`,
  currentPageTooSmall: '"currentPage" cannot be less than 1.',
  currentPageTooBig: '"currentPage" cannot be larger than "totalPages".',
  mustUsePageAsPathParam: `Your path parameter must be "[page]". See https://nextjs.org/docs/routing/dynamic-routes for more`,
};

const getPagination = (currentPage, totalPages) => {
  // maximum length of the Pagination Bar, should be an odd integer, default is 11
  const MAX_VISIBLE_ELEMENTS = 11;
  const ELEMENTS_ON_ONE_SIDE = (MAX_VISIBLE_ELEMENTS - 1) / 2; // 5

  const isTruncatingRequired = totalPages >= MAX_VISIBLE_ELEMENTS;

  const isLeftSideLengthy = currentPage - 1 > ELEMENTS_ON_ONE_SIDE;
  const isRightSideLengthy = totalPages - currentPage > ELEMENTS_ON_ONE_SIDE;

  const shouldTruncateStart = isTruncatingRequired && isLeftSideLengthy;
  const shouldTruncateEnd = isTruncatingRequired && isRightSideLengthy;

  const truncateStartOnly = shouldTruncateStart && !shouldTruncateEnd;
  const truncateEndOnly = !shouldTruncateStart && shouldTruncateEnd;

  let paginationStart;
  let paginationEnd;

  if (truncateStartOnly) {
    paginationStart = totalPages - MAX_VISIBLE_ELEMENTS + 3;
    paginationEnd = totalPages;
  } else if (truncateEndOnly) {
    paginationStart = 1;
    paginationEnd = MAX_VISIBLE_ELEMENTS - 2;
  } else {
    paginationStart = shouldTruncateStart ? currentPage - (ELEMENTS_ON_ONE_SIDE - 2) : 1;
    paginationEnd = shouldTruncateEnd ? currentPage + (ELEMENTS_ON_ONE_SIDE - 2) : totalPages;
  }

  const paginationLength = paginationEnd - paginationStart + 1;

  return {
    paginationStart,
    paginationLength,
    shouldTruncateStart,
    shouldTruncateEnd,
  };
};

// eslint-disable-next-line react/prop-types
const PaginationItems = ({ currentPage, pathname, query, totalPages }) => {
  const {
    paginationStart,
    paginationLength,
    shouldTruncateStart,
    shouldTruncateEnd,
  } = getPagination(currentPage, totalPages);

  const PaginationItemArray = [...Array(paginationLength)].map((_, index) => {
    const page = index + paginationStart;
    const isCurrent = page === currentPage;

    return (
      <PaginationItem
        key={page}
        value={page}
        isCurrent={isCurrent}
        testId={`page ${page}`}
        pathname={pathname}
        query={query}
      >
        {page}
      </PaginationItem>
    );
  });

  return (
    <>
      {shouldTruncateStart && (
        <>
          <PaginationItem key="1" value={1} query={query} testId="page 1" pathname={pathname}>
            1
          </PaginationItem>

          <PaginationItem key="separatorStart" testId="separatorStart" pathname={pathname}>
            &hellip;
          </PaginationItem>
        </>
      )}

      {PaginationItemArray}

      {shouldTruncateEnd && (
        <>
          <PaginationItem key="separatorEnd" testId="separatorEnd" pathname={pathname}>
            &hellip;
          </PaginationItem>

          <PaginationItem
            key={totalPages}
            value={totalPages}
            query={query}
            testId={`page ${totalPages}`}
            pathname={pathname}
          >
            {totalPages}
          </PaginationItem>
        </>
      )}
    </>
  );
};

function Pagination({ currentPage, pathname, query, totalPages }) {
  /* Developer Errors */
  if (process.env.NODE_ENV !== 'production') {
    const isCurrentPageTooSmall = currentPage < 1;

    if (isCurrentPageTooSmall) {
      const errorMessage = `${developmentErrors.currentPageValue(currentPage)} ${
        developmentErrors.currentPageTooSmall
      }`;

      throw new Error(errorMessage);
    }

    const isCurrentPageTooBig = currentPage > totalPages;
    if (isCurrentPageTooBig) {
      const errorMessage = `${developmentErrors.currentPageValue(currentPage)} ${
        developmentErrors.currentPageTooBig
      }`;

      throw new Error(errorMessage);
    }

    if (!pathname.endsWith('[page]')) {
      throw new Error(developmentErrors.mustUsePageAsPathParam);
    }
  }

  return (
    <nav className={styles.Pagination} data-testid="Pagination">
      <ol>
        {currentPage > 1 && (
          <PaginationItem
            value={currentPage - 1}
            pathname={pathname}
            query={query}
            testId={PREV_PAGE_BUTTON}
          >
            <LeftAngleIcon className={styles.icon} />
          </PaginationItem>
        )}

        <PaginationItems
          currentPage={currentPage}
          totalPages={totalPages}
          pathname={pathname}
          query={query}
        />

        {currentPage < totalPages && (
          <PaginationItem
            value={currentPage + 1}
            pathname={pathname}
            query={query}
            testId={NEXT_PAGE_BUTTON}
          >
            <RightAngleIcon className={styles.icon} />
          </PaginationItem>
        )}
      </ol>
    </nav>
  );
}

export default Pagination;
