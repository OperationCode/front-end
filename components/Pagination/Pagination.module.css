.Pagination > ol {
  display: flex;
  margin: 0 auto;
  padding: 0;
  justify-content: space-between;
  align-items: center;
  max-width: fit-content;
}

.icon {
  width: 100%;
}
