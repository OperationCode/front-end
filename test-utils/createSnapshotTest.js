/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
