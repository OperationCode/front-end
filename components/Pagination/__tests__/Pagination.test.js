import React from 'react';
import { cleanup, render } from '@testing-library/react';
import createSnapshotTest from 'test-utils/createSnapshotTest';

import Pagination, { developmentErrors } from '../Pagination';

describe('Pagination', () => {
  const requiredProps = {
    currentPage: 1,
    totalPages: 20,
    pathname: '/resources/[page]',
    query: {},
  };

  afterEach(cleanup);

  it('should render with required props', () => {
    createSnapshotTest(<Pagination {...requiredProps} />);
  });

  it('should not have a separator at start if not needed', () => {
    const { queryByTestId } = render(
      <Pagination {...requiredProps} currentPage={6} totalPages={11} />,
    );

    expect(queryByTestId('Pagination Item - separatorStart')).not.toBeInTheDocument();
  });

  it('should not have a separator at end if not needed', () => {
    const { queryByTestId } = render(
      <Pagination {...requiredProps} currentPage={6} totalPages={11} />,
    );

    expect(queryByTestId('Pagination Item - separatorEnd')).not.toBeInTheDocument();
  });

  it('should have a separator at start if needed', () => {
    const { queryByTestId } = render(
      <Pagination {...requiredProps} currentPage={7} totalPages={11} />,
    );

    expect(queryByTestId('separatorStart')).toBeInTheDocument();
  });

  it('should have a separator at end if needed', () => {
    const { queryByTestId } = render(
      <Pagination {...requiredProps} currentPage={5} totalPages={11} />,
    );

    expect(queryByTestId('separatorEnd')).toBeInTheDocument();
  });

  it('should throw an error if given value of currentPage is less than 1', () => {
    /* eslint-disable function-paren-newline */
    expect(() =>
      Pagination({ currentPage: 0, totalPages: 5, pathname: requiredProps.pathname }),
    ).toThrow(developmentErrors.currentPageTooSmall);
    /* eslint-enable function-paren-newline */
  });

  it('should throw an error if given value of currentPage is greater than totalPages', () => {
    /* eslint-disable function-paren-newline */
    expect(() =>
      Pagination({ currentPage: 6, totalPages: 5, pathname: requiredProps.pathname }),
    ).toThrow(developmentErrors.currentPageTooBig);
    /* eslint-enable function-paren-newline */
  });

  it('should throw an error if not using `[page]` in route path param', () => {
    /* eslint-disable function-paren-newline */
    expect(() => Pagination({ ...requiredProps, pathname: '/resources' })).toThrow(
      developmentErrors.mustUsePageAsPathParam,
    );
    /* eslint-enable function-paren-newline */
  });
});
