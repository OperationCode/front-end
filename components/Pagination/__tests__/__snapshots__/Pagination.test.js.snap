// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render with required props 1`] = `
<nav
  className="Pagination"
  data-testid="Pagination"
>
  <ol>
    <li
      className="PaginationItem current"
      data-testid="page 1"
    >
      <a
        className="unstyledLink"
        href="/resources/1"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        1
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 2"
    >
      <a
        className="unstyledLink"
        href="/resources/2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        2
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 3"
    >
      <a
        className="unstyledLink"
        href="/resources/3"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        3
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 4"
    >
      <a
        className="unstyledLink"
        href="/resources/4"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        4
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 5"
    >
      <a
        className="unstyledLink"
        href="/resources/5"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        5
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 6"
    >
      <a
        className="unstyledLink"
        href="/resources/6"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        6
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 7"
    >
      <a
        className="unstyledLink"
        href="/resources/7"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        7
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 8"
    >
      <a
        className="unstyledLink"
        href="/resources/8"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        8
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="page 9"
    >
      <a
        className="unstyledLink"
        href="/resources/9"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        9
      </a>
    </li>
    <li
      className="PaginationItem notClickable"
      data-testid="separatorEnd"
    >
      â€¦
    </li>
    <li
      className="PaginationItem"
      data-testid="page 20"
    >
      <a
        className="unstyledLink"
        href="/resources/20"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        20
      </a>
    </li>
    <li
      className="PaginationItem"
      data-testid="NEXT_PAGE_BUTTON"
    >
      <a
        className="unstyledLink"
        href="/resources/2"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        <span
          className="ScreenReaderOnly"
          data-testid="SCREEN_READER_ONLY"
        >
          Go to page
        </span>
        <svg
          className="icon"
        />
      </a>
    </li>
  </ol>
</nav>
`;
